src/shortestpaths/dijkstra.jl

```diff
@@ -37,59 +37,59 @@ function dijkstra_shortest_paths(g::AbstractGraph,
 
     nvg = nv(g)
     dists = fill(typemax(T), nvg)
+    dists[srcs] .= zero(T)
     parents = zeros(U, nvg)
-    preds = fill(Vector{U}(), nvg)
     visited = zeros(Bool, nvg)
+    visited[srcs] .= true
+
     pathcounts = zeros(Int, nvg)
-    H = PriorityQueue{U,T}()
-    dists[srcs] .= zero(T)
     pathcounts[srcs] .= 1
+    preds = fill(Vector{U}(), nvg)
+    # fill creates only one array.
 
     closest_vertices = Vector{U}()  # Maintains vertices in order of distances from source
-
     sizehint!(closest_vertices, nvg)
 
+    H = PriorityQueue{U,T}()
     for v in srcs
         H[v] = dists[v]
-        visited[v] = true
     end
 
     while !isempty(H)
-        hentry = dequeue_pair!(H)
-            # info("Popped H - got $(hentry.vertex)")
-        u = hentry[1]
+        u = dequeue!(H)
 
         if trackvertices
             push!(closest_vertices, u)
         end
 
+        d = dists[u] # Cannot be typemax if `u` is in the queue
         for v in outneighbors(g, u)
-            alt = (dists[u] == typemax(T)) ? typemax(T) : dists[u] + distmx[u, v]
+            alt = d + distmx[u, v]
 
             if !visited[v]
+                visited[v] = true
                 dists[v] = alt
                 parents[v] = u
+
                 pathcounts[v] += pathcounts[u]
-                visited[v] = true
                 if allpaths
                     preds[v] = [u;]
                 end
                 H[v] = alt
-                # info("Pushed $v")
-            else
-                if alt < dists[v]
-                    dists[v] = alt
-                    parents[v] = u
-                    #615
-                    pathcounts[v] = 0
-                    preds[v] = []
-                    H[v] = alt
+            elseif alt < dists[v]
+                dists[v] = alt
+                parents[v] = u
+                #615
+                pathcounts[v] = pathcounts[u]
```

Comment:
if `alt < dists[v]` then : for every path `P` ending at `u`, there is one path `[P, v]` that ends at `v`.

So `pathcounts[v] = pathcounts[u]` should be correct.

Reply:
@SohamTamba Can you give a case on which it fails?

Reply:
If I replace L83 with `pathcounts[v] = 0` then in `tests/dijkstra.jl`, the test on line 78 fails.
But that aside,
```
g = Cycle(3)
distsmx[1, 2]= 1
distmx[2, 3] = 1
distmx[1, 3] = 100
src = 1
```
Will result in `pathcounts = [1, 1, 0]` when it should be `pathcounts = [1, 1, 1]`.

Reply:
I don't understand why this changes in your code, when the previous version is correct:
```
julia> z = dijkstra_shortest_paths(g, 1)
LightGraphs.DijkstraState{Int64,Int64}([0, 1, 1], [0, 1, 1], Array{Int64,1}[Int64[], Int64[], Int64[]], [1, 1, 1], Int64[])

julia> z.pathcounts
3-element Array{Int64,1}:
 1
 1
 1
```

Reply:
@sbromberger It is required as we have added `elseif`in line 89.

Reply:
It is because of the first comment in @somil55 review to reduce the number of if else conditions.

Reply:
It was a bit confusing. So I told Soham to change.

Reply:
I think this is more confusing, frankly. Before, the `if / else` were complete: that is, which branch you took was evident from the `if` condition. Within the `else`, there is another branching, but it is incomplete (that is, it doesn't do anything when `alt > dists[v]`). It's still legible, though.

Did we change this because of a performance improvement?

Reply:
To be clear: I am interested in discussion, NOT a reversion to what we had before (which is what I currently prefer). Let's talk about it.

Reply:
```
if alt < dists[v]
        dists[v] = alt
        parents[v] = u
        #615
        pathcounts[v] = 0
        preds[v] = []
        H[v] = alt
end
if alt == dists[v]
        pathcounts[v] += pathcounts[u]
        if allpaths
                 push!(preds[v], u)
        end
end
```
In above snippet whenever `alt < dists[v]` branch is taken `alt==dists[v]` branch is also taken. This makes it confusing. I we change it to `elseif` then the branches are handling all three possible cases explicitly.

Reply:
Yes. That's the key, isn't it? It was so confusing it even got me confused yesterday. I'm glad you changed it, and I'm ok now.